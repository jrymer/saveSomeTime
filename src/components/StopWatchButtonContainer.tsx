import React from 'react';
import { connect } from 'react-redux';
import { View, Text, StyleSheet } from 'react-native';
import { setStopwatchTime } from '../store/stopwatchState/stopwatch.actions';
import { ISaveSomeTimeState, IStopwatchState } from '../models/saveSomeTime';
import { button, theme, text } from '../styles';
import { AppButton } from './AppButton';

interface IStopWatchtimerButtonsContainerState {
  milliseconds: number;
  startTime: number;
  isStopwatchRunning: boolean;
}

interface IStopWatchtimerButtonsContainerProps {
  stopwatchState: IStopwatchState,
  setStopwatchTime: typeof setStopwatchTime
}

class StopWatchtimerButtonsContainer extends React.Component<IStopWatchtimerButtonsContainerProps, IStopWatchtimerButtonsContainerState> {
  private interval = 0;

  constructor(props: IStopWatchtimerButtonsContainerProps) {
    super(props);

    this.state = {
      milliseconds: 0,
      startTime: 0,
      isStopwatchRunning: false
    };
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  private tick = () => {
    const difference = new Date().getTime() - this.state.startTime;
    this.setState((prevState: IStopWatchtimerButtonsContainerState) => ({ ...prevState, milliseconds: difference }));
  }

  private startStopwatch = () => {
    let { startTime } = this.state;

    if (startTime === 0) {
      startTime = new Date().getTime();
    }

    this.setState((prevState: IStopWatchtimerButtonsContainerState) => (
      {
        ...prevState,
        isStopwatchRunning: true,
        startTime
      }));

    this.interval = setInterval(
      () => this.tick(),
      10);
  }

  private stopStopwatch = () => {
    this.setState((prevState: IStopWatchtimerButtonsContainerState) => (
      {
        ...prevState,
        isStopwatchRunning: !prevState.isStopwatchRunning
      }));
    clearInterval(this.interval);
  }

  private resetStopwatch = () => {
    this.setState((prevState: IStopWatchtimerButtonsContainerState) => (
      { ...prevState, isStopwatchRunning: false, milliseconds: 0, startTime: 0 }
    ));
    clearInterval(this.interval);
  };

  private getDisplayValue = (milliseconds: number) => {
    let seconds = milliseconds / 1000;
    let hours = seconds / 3600; // 3,600 seconds in 1 hour
    seconds = seconds % 3600; // seconds remaining after extracting hours
    let minutes = seconds / 60; // 60 seconds in 1 minute
    seconds = seconds % 60;

    const secondsString = this.getLeadingZero(seconds);
    const minutesString = this.getLeadingZero(minutes);
    const hoursString = this.getLeadingZero(hours);
    const timeString = `${hoursString}:${minutesString}:${secondsString}`;
    return timeString;
  }

  private getLeadingZero = (value: number): string => {
    const time = Math.floor(value);

    return time < 10
      ? `0${time}`
      : time.toString();
  }

  private save = () => {
    this.props.setStopwatchTime(this.state.milliseconds);
  }

  private getSaveButton = () => {
    return (!this.state.isStopwatchRunning && this.state.milliseconds > 0)
    ? <AppButton buttonStyles={styles.saveEnabled} textStyles={text.whiteText} onPress={this.save} label="Save"></AppButton>
    : <AppButton buttonStyles={styles.saveDisabled} textStyles={text.whiteText} label="Save" disabled={true}></AppButton>;
  }

  private getStopwatchButton = () => {
    return !this.state.isStopwatchRunning
      ? <AppButton buttonStyles={styles.startButton} textStyles={text.whiteText} onPress={this.startStopwatch} label="Start"></AppButton>
      : <AppButton buttonStyles={styles.stopButton} textStyles={text.whiteText} onPress={this.stopStopwatch} label="Stop"></AppButton>
  }

  render() {
    return (
      <View style={styles.componentContainer}>
        <Text style={styles.timer}>
          {this.getDisplayValue(this.state.milliseconds)}
        </Text>
        <View style={styles.timerButtonsContainer}>
          <AppButton buttonStyles={styles.resetButton} textStyles={text.whiteText} onPress={this.resetStopwatch} label="Reset"></AppButton>
          {this.getStopwatchButton()}
        </View>
        {this.getSaveButton()}
      </View>
    )
  }
}

const styles = StyleSheet.create({
  componentContainer: {
    flexDirection: 'column',
    alignItems: 'center'
  },
  startButton: {
    ...button.roundButton,
    backgroundColor: theme.primary,
    width: 100,
    height: 40
  },
  stopButton: {
    ...button.roundButton,
    backgroundColor: theme.accent,
    width: 100,
    height: 40
  },
  resetButton: {
    ...button.roundButton,
    backgroundColor: theme.warning,
    width: 100,
    height: 40
  },
  saveDisabled: {
    ...button.roundButton,
    backgroundColor: theme.darkLighter,
    width: 210,
    height: 40,
    marginTop: 10
  },
  saveEnabled: {
    ...button.roundButton,
    backgroundColor: theme.dark,
    width: 210,
    height: 40,
    marginTop: 10
  },
  timerButtonsContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    width: 210
  },
  timer: {
    fontSize: 60,
    textAlign: 'center'
  }
});

const mapStateToProps = (state: ISaveSomeTimeState) => {
  const { stopwatchState } = state;
  return { stopwatchState };
};

const mapDispatchToProps = { setStopwatchTime };

export default connect(mapStateToProps, mapDispatchToProps)(StopWatchtimerButtonsContainer);
