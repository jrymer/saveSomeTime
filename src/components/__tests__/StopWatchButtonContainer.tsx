import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import { setStopwatchTime } from '../../store/stopwatchState/stopwatch.actions';
import StopWatchButtonContainer from '../StopWatchButtonContainer';

const initialState = {}; // here it is possible to pass in any middleware if needed into //configureStore
const mockStore = configureStore();

const props = {
  stopwatchState: {
    stopwatchRunning: false,
    stopwatchTime: 1000
  },
  setStopwatchTime
};

describe('StopWatchButtonContainer Tests', () => {
  it('StopWatchButtonContainer correctly w/ snapshot', () => {
    const store = mockStore(initialState);
    const tree = renderer.create(
      <Provider store={store}>
        <StopWatchButtonContainer {...props}
        />
      </Provider>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
