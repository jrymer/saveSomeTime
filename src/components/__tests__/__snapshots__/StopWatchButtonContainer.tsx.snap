// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StopWatchButtonContainer Tests StopWatchButtonContainer correctly w/ snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flexDirection": "column",
    }
  }
>
  <Text
    style={
      Object {
        "fontSize": 60,
        "textAlign": "center",
      }
    }
  >
    00:00:00
  </Text>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "width": 210,
      }
    }
  >
    <View
      accessible={true}
      clickable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#ce6a6b",
          "borderRadius": 20,
          "flexDirection": "row",
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "width": 100,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFF",
          }
        }
      >
        Reset
      </Text>
    </View>
    <View
      accessible={true}
      clickable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#4a919e",
          "borderRadius": 20,
          "flexDirection": "row",
          "height": 40,
          "justifyContent": "center",
          "opacity": 1,
          "width": 100,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#FFF",
          }
        }
      >
        Start
      </Text>
    </View>
  </View>
  <View
    accessible={true}
    clickable={true}
    isTVSelectable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#9096A9",
        "borderRadius": 20,
        "flexDirection": "row",
        "height": 40,
        "justifyContent": "center",
        "marginTop": 10,
        "opacity": 1,
        "width": 210,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#FFF",
        }
      }
    >
      Save
    </Text>
  </View>
</View>
`;
